File: ui/src/types/portfolio.ts
export interface Metadata {
  name?: string
  generateName?: string
  creationTimestamp?: string
  deletionTimestamp?: string
  version?: number
}

export interface ExtensionList<T> {
  page: number
  size: number
  total: number
  items: T[]
  first: boolean
  last: boolean
  hasNext: boolean
  hasPrevious: boolean
  totalPages: number
}

export interface PortfolioSpec {
  displayName: string
  slug: string
  description?: string
  cover?: string
  publicView?: boolean
  priority?: number
}

export interface PortfolioStatus {
  projectCount?: number
}

export interface Portfolio {
  apiVersion: 'portfolio.plugin.halo.run/v1alpha1'
  kind: 'Portfolio'
  metadata: Metadata
  spec: PortfolioSpec
  status?: PortfolioStatus
}

export type PortfolioOptionType = 'TECH_STACK' | 'SOURCE' | 'DOMAIN'

export interface PortfolioOptionSpec {
  portfolioName: string
  type: PortfolioOptionType
  value: string
  label: string
  sortOrder?: number
}

export interface PortfolioOption {
  apiVersion: 'portfolio.plugin.halo.run/v1alpha1'
  kind: 'PortfolioOption'
  metadata: Metadata
  spec: PortfolioOptionSpec
}

export interface PortfolioProjectSpec {
  portfolioName: string
  title: string
  summary?: string
  description?: string
  postName?: string
  coverImage?: string
  gallery?: string[]
  tags?: string[]
  domain?: string
  techStack?: string[]
  source?: string
  sourceDetail?: string
  featured?: boolean
  priority?: number
  startDate?: string
  endDate?: string
  published?: boolean
}

export interface PortfolioProject {
  apiVersion: 'portfolio.plugin.halo.run/v1alpha1'
  kind: 'PortfolioProject'
  metadata: Metadata
  spec: PortfolioProjectSpec
}

export const API_GROUP = '/apis/portfolio.plugin.halo.run/v1alpha1'
export const PORTFOLIOS_API = `${API_GROUP}/portfolios`
export const PORTFOLIO_OPTIONS_API = `${API_GROUP}/portfoliooptions`
export const PORTFOLIO_PROJECTS_API = `${API_GROUP}/portfolioprojects`

File: src/main/java/run/halo/portfolio/extension/Portfolio.java
package run.halo.portfolio.extension;

import static io.swagger.v3.oas.annotations.media.Schema.RequiredMode.REQUIRED;

import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import run.halo.app.extension.AbstractExtension;
import run.halo.app.extension.GVK;

@Data
@ToString(callSuper = true)
@EqualsAndHashCode(callSuper = true)
@GVK(group = "portfolio.plugin.halo.run", version = "v1alpha1", kind = "Portfolio",
    plural = "portfolios", singular = "portfolio")
public class Portfolio extends AbstractExtension {

    @Schema(requiredMode = REQUIRED)
    private PortfolioSpec spec;

    @Schema
    private PortfolioStatus status;

    @Data
    public static class PortfolioSpec {
        @Schema(requiredMode = REQUIRED, minLength = 1)
        private String displayName;

        @Schema(requiredMode = REQUIRED, pattern = "^[a-z0-9]([-a-z0-9]*[a-z0-9])?$")
        private String slug;

        @Schema
        private String description;

        @Schema
        private String cover;

        @Schema(defaultValue = "true")
        private Boolean publicView;

        @Schema(defaultValue = "0")
        private Integer priority;
    }

    @Data
    public static class PortfolioStatus {
        @Schema(defaultValue = "0")
        private Integer projectCount;

        @Schema(description = "True when another portfolio uses the same slug")
        private Boolean slugConflict;
    }
}
